### PR TITLE
Integrate Tendermint consensus into Eudico

### PR DESCRIPTION
**Pull request by dnkolegov**
## Proposed Changes
This PR integrates the Tendermint consensus protocol into Eudico.

## Additional Info
[Tendermint as a Consensus Protocol in Eudico](https://hackmd.io/@consensuslab/SJg-BGBeq).

### CODE REVIEW

File: itests/kit/docker/docker.go
Comment (adlrocha) [inline itests/kit/docker/docker.go:36]: I don't think we should worry about this now, but this may fail for people that require sudo privileges to run docker :)

File: scripts/eud-tendermint-node2.sh
Comment (adlrocha) [inline scripts/eud-tendermint-node2.sh:1]: Can we add all Tendermint-specific scripts into the same folder including maybe a quick explanation of what each of them do? Thanks!

**Review by adlrocha — CHANGES_REQUESTED** [review]
It look really good. There is one critical thing to fix before merging in `tspow/mine.go` where we've removed support for cross-messages. Apart from that feel free to fix my comments at your discretion. If you see any of them are not a priority add them as an issue for the future.

(I would also suggest merging this PR after rebasing with atomic execution protocol just in case if you don't mind, but whatever works best for you).

File: chain/consensus/tspow/mine.go
Comment (dnkolegov) [reply chain/consensus/tspow/mine.go:112]: Sure

File: chain/consensus/tspow/mine.go
Comment (dnkolegov) [reply chain/consensus/tspow/mine.go:84]: I forgot to uncomment this line after debugging.

File: chain/consensus/tendermint/tendermint.go
Comment (dnkolegov) [reply chain/consensus/tendermint/tendermint.go:322]: Created an issue https://github.com/filecoin-project/eudico/issues/142 to address that.

File: chain/consensus/tendermint/tendermint.go
Comment (dnkolegov) [reply chain/consensus/tendermint/tendermint.go:331]: Created an issue https://github.com/filecoin-project/eudico/issues/142 to address that.

File: chain/consensus/common/cns_validations.go
Comment (dnkolegov) [reply chain/consensus/common/cns_validations.go:421]: Fixed

File: chain/consensus/common/cns_validations.go
Comment (dnkolegov) [reply chain/consensus/common/cns_validations.go:391]: That's correct. But these functions have a valuable difference.
The original function just checks the validity of the block messages.
The new function filters bad messages out of the block.
So I suggest refactoring the code if Tendermint is in production or at least lives long-time.

File: chain/consensus/tendermint/app.go
Comment (dnkolegov) [reply chain/consensus/tendermint/app.go:121]: Yes, you are right. We register a Tendermint blockchain offset for a new subnet here.
Suppose we just started a root subnet and Tendermint blockchain, then its offset will be for example 3 blocks.
And then after an hour, we decided to register a new subnet, its offset will be 1000. If we don't use offsets then the second subnet will have to access all previous Tendermint blocks. 

> Why I ask is because I am wondering how critical would it be for a rogue node to register a subnet he is not part of (or that doesn't even exist yet).
It is critical. A rogue node can register a network earlier than it would be registered, e.g. at epoch h. Then when the network is instantiated by real participants at epoch H they will have to process all blocks from epoch h to H.

File: chain/consensus/common/cns_validations.go
Comment (adlrocha) [reply chain/consensus/common/cns_validations.go:391]: Sounds good. Add a quick note so we remember, but apart from that, I am happy with it.

**Comment by adlrocha** [comment]
I see all the critical issues have been fixed. The two big outstanding ones would be the script organization, and the use of CID instead of a deterministic hash to identify messages and blocks, but as they are not critical I'll approve the PR so you can merge when you are ready. 

Feel free to ping me if you need me to do another pass before merging.

File: chain/consensus/tendermint/utils.go
Comment (dnkolegov) [reply chain/consensus/tendermint/utils.go:173]: This code is used once to get information about the Tendermint node key and address.

File: chain/consensus/tendermint/app.go
Comment (dnkolegov) [reply chain/consensus/tendermint/app.go]: Fixed. Now only one subnet can be registered in Tendermint. It doesn't protect against DoS attacks but mitigates them.

File: chain/consensus/tendermint/message_pool.go
Comment (dnkolegov) [reply chain/consensus/tendermint/message_pool.go]: Fixed

File: chain/consensus/tendermint/message_pool.go
Comment (dnkolegov) [reply chain/consensus/tendermint/message_pool.go]: Fixed

File: chain/consensus/tendermint/mine.go
Comment (dnkolegov) [reply chain/consensus/tendermint/mine.go]: Fixed

File: chain/consensus/tendermint/mine.go
Comment (dnkolegov) [reply chain/consensus/tendermint/mine.go]: Fixed

File: scripts/eud-tendermint-node2.sh
Comment (dnkolegov) [reply scripts/eud-tendermint-node2.sh]: Fixed

File: chain/consensus/tendermint/message_cache.go
Comment (dnkolegov) [reply chain/consensus/tendermint/message_cache.go:44]: https://github.com/filecoin-project/eudico/blob/tendermint-exp/chain/consensus/tendermint/mine.go#L143

File: chain/consensus/tendermint/mine.go
Comment (dnkolegov) [reply chain/consensus/tendermint/mine.go:142]: I would prefer to address this later. Because it will complicate the implementation. The cache is just a map now and for a production-ready system it should be something else, which means that we will have to rewrite it anyway. So let's defer for the future.

File: chain/consensus/tendermint/tendermint.go
Comment (dnkolegov) [reply chain/consensus/tendermint/tendermint.go]: tag was removed

Comment (dnkolegov) [reply chain/consensus/tendermint/tendermint.go]: removed

File: chain/consensus/tendermint/mine.go
Comment (adlrocha) [reply chain/consensus/tendermint/mine.go:142]: Can you add a comment (with NOTE or FIXME) so it is easy to locate in the future? Thanks!

File: chain/consensus/tendermint/mine.go
Comment (adlrocha) [inline chain/consensus/tendermint/mine.go:170]: Maybe make this log a `WARN`?

File: chain/consensus/tendermint/README.md
Comment (adlrocha) [inline chain/consensus/tendermint/README.md:194]: I love this!! We should maybe make it the baseline tests any consensus implementation needs to at least pass (maybe we can add something else, but as a baseline it is perfect)

File: cmd/eudico/benchmark.go
Comment (adlrocha) [inline cmd/eudico/benchmark.go:27]: This length determines the number of epochs you will be monitoring the node for the benchmark, right?

**Review by adlrocha — APPROVED** [review]
LGTM!

File: cmd/eudico/benchmark.go
Comment (dnkolegov) [reply cmd/eudico/benchmark.go:27]: Right

File: chain/consensus/tendermint/mine.go
Comment (dnkolegov) [reply chain/consensus/tendermint/mine.go:170]: fixed

File: itests/kit/docker/docker.go
Comment (dnkolegov) [reply itests/kit/docker/docker.go]: ok

**Pull request merged**
